File: python/08_015.py
import heapq

from dataclasses import dataclass
from typing import List, Tuple

@dataclass
class Request:
   id: int
   size: int

@dataclass
class Server:
   id: int
   capacity: int
   current_load: int = 0
   last_active: int = 0

def assign_requests(requests: List[Request], 
                  servers: List[Server]) -> List[Tuple[int, int]]:
   server_heap = [(0, 0, server.id) for server in servers]
   heapq.heapify(server_heap)
   
   server_map = {server.id: server for server in servers}
   
   assignments = []
   current_time = 0
   
   for request in requests:
      current_time += 1
      
      if not server_heap:
         break
         
      load, last_active, server_id = heapq.heappop(
         server_heap)
      server = server_map[server_id]
      
      if server.current_load + request.size <= server.capacity:
         server.current_load += request.size
         server.last_active = current_time
         assignments.append((server_id, request.id))
         
         heapq.heappush(server_heap, 
                     (server.current_load, -current_time, 
                     server_id))
   
   return assignments

File: cpp/08_015.cpp
#include <vector>
#include <queue>
#include <unordered_map>
#include <tuple>

struct Request {
    int id;
    int size;
    
    Request(int id, int size) : id(id), size(size) {}
};

struct Server {
    int id;
    int capacity;
    int currentLoad;
    int lastActive;
    
    Server(int id, int capacity) 
        : id(id), capacity(capacity), currentLoad(0), 
          lastActive(0) {}
};

// Custom comparator for priority queue
struct ServerComparator {
    bool operator()(
            const std::tuple<int, int, int>& a, 
            const std::tuple<int, int, int>& b) {
        // Compare by load first
        if (std::get<0>(a) != std::get<0>(b)) {
            return std::get<0>(a) > std::get<0>(b); // Min heap
        }
        // If loads are equal, compare by last active time
        return std::get<1>(a) > std::get<1>(b); // Min heap
    }
};

std::vector<std::pair<int, int>> assignRequests(
        const std::vector<Request>& requests, 
        std::vector<Server>& servers) {
    // Create a map from server ID to Server object
    std::unordered_map<int, Server*> serverMap;
    for (auto& server : servers) {
        serverMap[server.id] = &server;
    }
    
    // Priority queue for servers: (load, lastActive, serverId)
    std::priority_queue
        std::tuple<int, int, int>, 
        std::vector<std::tuple<int, int, int>>, 
        ServerComparator> serverHeap;
    
    for (const auto& server : servers) {
        serverHeap.push(
            std::make_tuple(0, 0, server.id));
    }
    
    std::vector<std::pair<int, int>> assignments;
    int currentTime = 0;
    
    for (const auto& request : requests) {
        currentTime++;
        
        // Get least loaded server
        if (serverHeap.empty()) {
            break; // No available servers
        }
        
        auto [load, lastActive, serverId] = serverHeap.top();
        serverHeap.pop();
        
        Server* server = serverMap[serverId];
        
        // Check if server has capacity for this request
        if (server->currentLoad + request.size <= 
            server->capacity) {
            // Assign request to server
            server->currentLoad += request.size;
            server->lastActive = currentTime;
            assignments.push_back({server->id, request.id});
            
            // Put server back in heap with updated load
            serverHeap.push(std::make_tuple(
                server->currentLoad, 
                -currentTime, 
                server->id));
        }
    }
    
    return assignments;
}